### PR TITLE
feat(desktop): SSRF hardening for the browser agent + review fixes

### PR DESCRIPTION
**Pull request by waleedlatif1**
## What

Security hardening for the new desktop Electron app, plus the shared-package extraction and review fixes that came out of a deep review of the `dev` electron branch. Targets **`dev`** (the branch the desktop app lives on).

### Shared SSRF primitive (`@sim/security/ssrf`)
Extracts IP classification into one audited, `ipaddr.js`-based module — `isPrivateIp` (resolved IP, fail-closed), `isPrivateIpHost` (host literal, fail-open on DNS names), `isIpLiteral`, `unwrapIpv6Brackets`. This **de-duplicates** three previously-divergent implementations:
- the fuller `isPrivateOrReservedIP` in `input-validation.server.ts`
- a weaker copy in `input-validation.ts` (missing the IPv4-compatible-IPv6 embedded check)
- ad-hoc regex `isPrivateIPv4`/`isPrivateOrLoopbackIPv6` in `sap_concur`, `zoominfo`, and the `sap` contract

All callers (apps/sim outbound-fetch validation, MCP SSRF, DB-host validation, and the desktop shell) now share one primitive. The comprehensive edge-case matrix moved into the package test.

### Browser-agent SSRF guard
The embedded agent browser is model/tool-driven, so a navigation to an internal host would let the page be read back via the snapshot/read tools. Fix:
- The agent partition's `onBeforeRequest` is the **single authoritative choke point**: document navigations (top-level + iframes) get a full DNS-resolving check — covering **page-initiated** navigation the tool layer never sees (server redirects, link clicks, `location.href`, meta-refresh) — while subresources get a cheap synchronous literal-IP backstop.
- `browser_navigate` / `browser_open_tab` also validate up front so the model gets a clean error instead of a silent load failure.
- Blocks loopback, RFC1918, link-local / cloud-metadata (`169.254.169.254`), and obfuscated / IPv4-mapped forms; hostnames are resolved (DNS-rebinding aware).

### Other hardening
- **crashReporter** — local-only native minidumps (`uploadToServer:false`, no backend); dump dir recorded in the event log.
- **CSP fallback** — a minimal, non-drifting policy (`frame-ancestors 'self'; object-src 'none'; base-uri 'self'`) injected only when an app-origin document ships no CSP of its own.
- **window.open (MCP OAuth)** gated to `https` so a renderer-controlled frame name can't ride an `http` URL into an in-app window.
- **updater** — shared `configureAutoUpdater` re-asserts `channel`/`allowDowngrade` on both the launch and manual-check paths.
- **handoff** — the loopback listener tears down only after the CSRF state validates, not on any format-valid request (self-DoS guard).
- **cdp** — per-`WebContents` callbacks so a background tab's events reach its own driver.
- **config** — drop the unreachable bare `::1` loopback entry.
- **README** — correct the App Sandbox / security-scoped-bookmark note for the Developer ID build.

## Testing
- `@sim/security/ssrf`: 67 tests
- `apps/desktop`: 200 tests
- affected `apps/sim` suites (input-validation, domain-check, mock consumers): green
- type-check + biome clean across `apps/desktop`, `packages/security`, and touched `apps/sim` files

## Notes for review
- Two findings were **deliberately not changed** (verified against the flow): the permissive `idp-in-window` navigation default (tightening breaks same-window OAuth connect), and the cross-origin MCP popup (same-origin-only would break the real provider OAuth URL — gated to `https` instead).

### CODE REVIEW

File: apps/desktop/src/main/handoff.ts [hunk 1]
```diff
@@ -87,7 +87,10 @@ export function createHandoffManager(
       response
         .writeHead(200, { 'Content-Type': 'text/html; charset=utf-8' })
         .end(CALLBACK_RESPONSE_HTML)
-      stopLoopback()
+      // Do NOT tear the server down here: a request with a format-valid but
+      // wrong state must not kill the one-shot listener before the genuine
+      // callback arrives. `consume()` stops the server once the state actually
+      // validates (via clear()); the TTL timer is the backstop otherwise.
```
[the remark this answers is not in the document]
Comment (waleedlatif1) [reply apps/desktop/src/main/handoff.ts]: Fixed in 486f9181. The loopback handler now gates `onCallback` on a non-consuming, constant-time `matchesPendingState(state)` check, so a format-valid but wrong/expired state gets the generic 200 page and is otherwise ignored — it no longer reaches `handleCallback`/`consume`, so no spurious "sign-in failed" UI fires while the genuine callback is still pending. The one-shot listener still survives wrong-state requests (the self-DoS guard), and `consume()` re-validates + tears it down when the real state lands. Added a test asserting the wrong-state request does not fire the callback.

File: apps/desktop/src/main/browser-agent/session.ts [hunk 2]
```diff
@@ -68,6 +69,26 @@ function configureAgentPartition(ses: Session): void {
...
+      void checkAgentUrl(details.url).then((guard) => {
+        if (!guard.ok) {
+          logger.warn('Blocked agent document navigation to a private host')
+        }
+        callback({ cancel: !guard.ok })
+      })
+      return
+    }
+    callback({ cancel: isBlockedRequestUrl(details.url) })
+  })
```
[the remark this answers is not in the document]
Comment (waleedlatif1) [reply apps/desktop/src/main/browser-agent/session.ts:98]: Fixed in 486f9181. The document-path `checkAgentUrl(...).then(...)` now has a `.catch()` that fails closed — an unexpected rejection logs and calls `callback({ cancel: true })` instead of leaving the request suspended.

File: apps/desktop/src/main/updater.ts [hunk 3]
```diff
@@ -218,14 +233,13 @@ export function checkForUpdatesInteractive(deps: UpdaterDeps): void {
...
-    logger.error('Manual update check failed', { error })
+  const autoUpdater = configureAutoUpdater()
+  if (!autoUpdater) {
+    return
   }
+  void autoUpdater.checkForUpdates().then((result) => {
+    if (!result || result.updateInfo.version === app.getVersion()) {
+      void dialog.showMessageBox({ type: 'info', message: 'Sim is up to date' })
+    }
+  })
```
[the remark this answers is not in the document]
Comment (waleedlatif1) [reply apps/desktop/src/main/updater.ts]: Fixed in 486f9181. `checkForUpdatesInteractive` now chains a `.catch()` that logs the error and surfaces an error dialog, matching the launch path's error handling instead of silently dropping network/manifest/cert failures.

File: apps/desktop/src/main/browser-agent/session.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (waleedlatif1) [reply apps/desktop/src/main/browser-agent/session.ts:98]: This is the documented DNS-rebinding TOCTOU, and it's an intentional trade-off rather than an oversight. Fully closing it requires egress firewalling: Chromium re-resolves DNS in its own network stack, independent of our main-process check, so even DNS-checking every subresource can't prevent a rebind between our lookup and Chromium's connection.

What is covered: the readable exfiltration vector — document content surfaced through the read/snapshot tools — is DNS-checked at the document choke point (`mainFrame`/`subFrame`). Cross-origin subresource **responses are unreadable** under `webSecurity: true`/CORS, so a subresource to an internal host is at most a *blind* request, not a data-exfil path. Literal private-IP subresources (including the `169.254.169.254` metadata endpoint) are blocked synchronously by `isBlockedRequestUrl`. The remaining blind-hostname-subresource case is exactly the acknowledged limitation called out in the code comment, and matches how the rest of the codebase treats rebinding (IP pinning on the Node fetch path, egress controls beyond it).

**Comment by waleedlatif1** [comment]
@greptile

**Comment by waleedlatif1** [comment]
@cursor review

**Comment by waleedlatif1** [comment]
Pushed a follow-up cleanup wave (whole-desktop audit): shared `sleep`/`getErrorMessage` in driver.ts, dead-code removal (LocalFilesystemService.clear, launcher isVisible, decideStartRoute dead param), extracted `isLoopbackIp`/`isLoopbackHostname` into `@sim/security/ssrf` and migrated domain-check/s3/desktop-config off their local copies, plus an api-validation baseline reconcile for dev's already-merged speech routes (sanctioned allowlist + annotation; ratchets kept tight). Also hardened the handoff callback to tear the loopback down on a genuine state match (closes a refresh-race spurious-dialog). Full line-by-line audit + equivalence checks green; 305 tests pass across desktop/security/sim.

**Comment by waleedlatif1** [comment]
@greptile

**Comment by waleedlatif1** [comment]
@cursor review

File: apps/desktop/src/main/browser-agent/url-guard.ts [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+      return BLOCKED
+    }
+  } catch (error) {
+    logger.info('Agent navigation host did not resolve; letting the load fail naturally', {
+      host,
+      error: getErrorMessage(error),
+    })
+  }
+
+  return OK
```
[the remark this answers is not in the document]
Comment (waleedlatif1) [reply apps/desktop/src/main/browser-agent/url-guard.ts:70]: Fixed in a266ebab. `checkAgentUrl` now fails closed on DNS resolution failure — it returns a block instead of allowing the load. You're right that the fail-open disagreed with `validateUrlWithDNS`, and the risk is real here: Chromium resolves DNS in its own stack independent of our Node `dns.lookup`, so a host our resolver can't resolve could still reach a private address via Chromium's resolver. Now aligned with the codebase's fail-closed convention; updated the test and TSDoc accordingly.

**Comment by waleedlatif1** [comment]
@greptile

**Comment by waleedlatif1** [comment]
@cursor review
